primitives/hp-verifiers/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright 2024, Horizen Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![deny(missing_docs)]
+
+//! The traits and basic implementations for the verifier pallets based on `pallet-verifiers`
+//! parametric pallet.
+
+use codec::{Decode, Encode, EncodeLike};
+use scale_info::TypeInfo;
+use sp_core::MaxEncodedLen;
+pub use sp_std::borrow::Cow;
+use sp_std::fmt::Debug;
+use sp_weights::Weight;
+
+/// Define the minimum traits that proofs and public inputs should implement.
+pub trait Arg: Debug + Clone + PartialEq + Encode + Decode + TypeInfo + MaxEncodedLen {}
+impl<T: Debug + Clone + PartialEq + Encode + Decode + TypeInfo + MaxEncodedLen> Arg for T {}
+/// Define the minimum traits that verification keys should implement.
+pub trait VkArg: Arg + EncodeLike {}
+impl<T: Arg + EncodeLike> VkArg for T {}
+
+/// The verification error type
+#[derive(Debug, PartialEq)]
+pub enum VerifyError {
+    /// Provided data has not valid public inputs.
+    InvalidInput,
+    /// Provided data has not valid proof.
+    InvalidProofData,
+    /// Verify proof failed.
+    VerifyError,
+    /// Provided an invalid verification key.
+    InvalidVerificationKey,
+}
+
+/// The trait that characterize a verifier.
+pub trait Verifier: 'static {
+    /// The proof format type accepted by the verifier
+    type Proof: Arg;
+    /// The public inputs format
+    type Pubs: Arg;
+    /// The verification key format
+    type Vk: VkArg;
+
+    /// The context used to generate the statements hash.
+    fn hash_context_data() -> &'static [u8];
```

Comment:
This can be a constant perhaps ?

Reply:
No it's not... Function intraits cannot be declared as `const`